### PR TITLE
Application version handling.

### LINKED ISSUE #56605: [RFC] Add extensible version infrastructure
### Problem description

At present, there is a way that applications can get various versions of things in zephyhr - they can get the version of zephyr that is checked out, they can get the version of a board that is being compiled for, but from an application perspective there is no unified way of getting a version of the application itself and users that want to have such a system in their applications must do so out of tree. The main bootloader supported in zephyr is MCUboot, which supports a modified semantic versioning system of a.b.c+d (a = major, b = minor, c = patch, d = build) where each of these fields is numeric. To date in zephyr, all applications are signed with version 0.0.0+0 for signing is enabled in a build, it can be manually over-wrote by including the desired version in the MCUboot signing options Kconfig variable, but similarly as mentioned before, this is a string that is provided directly the imgtool command line option and cannot be easily used or accessed by applications.

Therefore the idea would be to have a new standardised version system in zephyr *for applications* that can be easily set or read from the build system, singing system and application code itself that can (in future) be extended or replaced with out of tree code so it can be used freely.

### Proposed change

I suggest that there be a new Kconfig added which specifies the version of the application, the format of this field would not be inherently restricted to any specific format but format restrictions could be applied depending upon the signing system used if it has strict requirements (i.e. imgtool does not support alphanumeric characters as part of a version). This version can default to `0.0.0+0` for semantic versioning and to stay constant with how things are set in zephyr already. A cmake file can then be used to generate an (optional) include file with the version string and also split the individual components up. The Kconfig itself can be used for e.g. imgtool to sign when the application is generated.

## Detailed RFC

Similar to the new support for external/out of tree replacement of the zephyr signing tasks, this system can be fully replaced if users so wish - as an example, a user might have a bootloader which takes just a single integer as the version, they can replace the cmake script which reads the Kconfig and generates the output file to customise it however they see fit, or supply it to their custom signing applications, by doing it like this and keeping it unlinked from the kernel version (and kernel version file) it allows complete flexibility for expansion by users without being limited to a set format.

The choice of Kconfig for the version is to allow for multiple different builds in the same application folder depending upon configuration or build type, it can also be set dynamically by the build system e.g. CI automatically incrementing the version when it performs a build. It would also allow the version of a zephyr application to be over-wrote without having to fork it, for example MCUboot could have a version set which users might want to change to their own scheme or increment the version because they've added a custom zephyr module/board/arch/soc without having changed the MCUboot base code/repository.

### Concerns and Unresolved Questions

How to get/store the version, if it should be inherently tied between an application version and the version provided to the imgtool/other tool application (if anyone can provide a valid reason/use case whereby they build an application and embed a version to it and the version that they use with imgtool when signing that application is **different** to the actual application version, I'd like to hear why. Simply saying "because" is not a valid reason as to why this would be a required feature)

## Alternatives

Let users do as they currently do and deal with everything themselves - not a good option

### PR DESCRIPTION
**Pull request by tejlmand**
This PR is inspired by #55814 proposal.

It tries to address the concerns raised in the review.

Main parts of this proposal :
- Zephyr version infrastructure is improved for better reuse, this allows a single infrastructure instead of duplicated files
  This means the same APP_VERSION_x variables similar to KERNEL_VERSION_x are available, providing consistency.
- APP_BUILD_VERSION is supported which includes extra build info through `git describe`, but executed in the application repo.
   For example:
   `#define KERNEL_BUILD_VERSION zephyr-v3.3.0-1081-gb6f97a492717`
   `#define APP_BUILD_VERSION v2.2.0-11-g9e229e382dd2`
- app VERSION file is placed in application folder and follows the semantics of Zephyr kernel VERSION file.
- A dedicated `MCUBOOT_IMGTOOL_SIGN_VERSION` defined, this value defaults to the APP_VERSION_STRING when application is in use, else it defaults to the old `0.0.0+0` string.

Fixes #56605

### CODE REVIEW

**Review by nordicjm — CHANGES_REQUESTED** [review]
Some issues with this:

- A dependency issue, modified hello_world to have a sample VERSION file then added `#include "app_version.h"` to main.c and built: `/tmp/aa/zephyr/samples/hello_world/src/main.c:8:10: fatal error: app_version.h: No such file or directory` - had to manually run `ninja app_version_h` to have this file generated
- Version that is generated is not valid for mcuboot, given the below VERSION file, signing then fails: `Error: Invalid value for '-v' / '--version': Invalid version number, should be maj.min.rev+build with later parts optional FATAL ERROR: command exited with status 2: /home/jamie/.local/bin/imgtool sign --version 1.2.3-5 --align 4 --header-size 512 --slot-size 483328 --key /tmp/aa/bootloader/mcuboot/root-rsa-2048.pem /tmp/aa/zephyr/samples/hello_world/_AA/zephyr/zephyr.bin /tmp/aa/zephyr/samples/hello_world/_AA/zephyr/zephyr.signed.bin`
- Changes fail to be picked up - if a project is configured and built, then VERSION is changed and you build again, the old version is used, the old version is also provided to the signing version Kconfig and actually you can't even force the new version to be propagated through manually, running `ninja app_version_h` just gives `ninja: no work to do.`
```
VERSION_MAJOR = 1
VERSION_MINOR = 2
PATCHLEVEL = 3
VERSION_TWEAK = 4
EXTRAVERSION = 5
```

This does not work well.

**Comment by nordicjm** [comment]
Also not sure how this would work with something like TF-M whereby there are multiple versions for different components in a single build - maybe @joerchan can see if this would work

**Comment by tejlmand** [comment]
> Also not sure how this would work with something like TF-M whereby there are multiple versions for different components in a single build

The application version is not enforced on to TF-M, and regardless of #56141 or #55814 approach it shouldn't.
What both approaches can offer is a default value when a version is active, but still allow users to individually specify TF-M version.

**Comment by nordicjm** [comment]
> > Also not sure how this would work with something like TF-M whereby there are multiple versions for different components in a single build
> 
> The application version is not enforced on to TF-M, and regardless of #56141 or #55814 approach it shouldn't. What both approaches can offer is a default value when a version is active, but still allow users to individually specify TF-M version.

Not the version of TF-M itself, the version of the secure core and non-secure core code that is being built, there are currently 2 Kconfigs for this, with the other PR this works fine because each image would get its own Kconfig, with this that is not possible.

> So if you used the same 1.2.3-5 as CONFIG_APPLICATION_VERSION in https://github.com/zephyrproject-rtos/zephyr/pull/55814, then the same would happen.

This generates the wrong version, the other PR does not, the other PR would have to have someone purposely ignore the Kconfig text and enter an invalid value.

**Comment by nordicjm** [comment]
And it looks like yours completely changes the kernel version string which breaks existing functionality from what the failing tests show

**Comment by tejlmand** [comment]
> And it looks like yours completely changes the kernel version string which breaks existing functionality from what the failing tests show

A `KERNEL` too much, so that it happened to become `KERNEL_BUILD_VERSION` instead of just `BUILD_VERSION`.
Fixed now.

**Comment by tejlmand** [comment]
> the old version is also provided to the signing version Kconfig and actually you can't even force the new version to be propagated 

This happens to be an existing bug in Zephyr, just raised an issue here:
https://github.com/zephyrproject-rtos/zephyr/issues/56235

and is exactly why I don't want to have multiple different infrastructures for version handling.
It increases the risk of a dependency issue fixed in one infrastructure is not applied for other infrastructures.

**Comment by tejlmand** [comment]
>  there are currently 2 Kconfigs for this, with the other PR this works fine because each image would get its own Kconfig, with this that is not possible.

This PR doesn't remove any existing options, meaning that TF-M still has their existing Kconfig settings, and they should continue to have so, cause how TF-M decides to do versioning is for TF-M project to decide, and Zephyr Kconfig interface to TF-M by providing identical settings.

This means users can continue to specify TF-M secure and non-secure version as they have always done.
What this PR provides, is an ability to automatically have a default based on Zephyr's app version if we wish so.

Same with the imgtool version, users are free to set the version value, and it is not forced to be the same for app and imgtool, although I personally would recommend to have them in sync.

This is exactly why imgtool should have its own Kconfig setting, as mentioned here: https://github.com/zephyrproject-rtos/zephyr/pull/55814#discussion_r1143003499
> if the intention is to be used for img tool, then the setting name should reflect this and the setting should be located with similar img tool setting.

**Comment by nordicjm** [comment]
So we go for a universal version but then have exclusions and have different systems that have their own version fields which are completely unrelated? That doesn't fix anything, it just brings confusion and a set of invisible rules.

**Comment by tejlmand** [comment]
> A dependency issue, modified hello_world to have a sample VERSION file then added ...

Added a dependency from `zephyr_interface`.
Altough Zephyr itself strictly speaking doesn't use this feature, then the app lib most likely will use it.
As well as any user specific Zephyr modules could want to use this header.
This ensures the `app_version.h` is always generated as part of the build.

> Version that is generated is not valid for mcuboot, given the below VERSION file, signing then fails

Fixed, default version for `MCUBOOT_IMGTOOL_SIGN_VERSION` now follows the `major.minor.patch` provided by the `VERSION` file, else it defaults to `0.0.0+0`.
https://github.com/zephyrproject-rtos/zephyr/blob/b30f828b9de1f0bf0f278f4d8e98f705fd0948cf/Kconfig.zephyr#L861-L864

**Review by nordicjm — DISMISSED** [review]

**Review by nordicjm — DISMISSED** [review]

**Review by nashif — DISMISSED** [review]

File: CMakeLists.txt [hunk 1]
```diff
@@ -544,12 +544,29 @@ add_custom_command(
...
+if(EXISTS ${APPLICATION_SOURCE_DIR}/VERSION)
+  add_custom_command(
+    OUTPUT ${PROJECT_BINARY_DIR}/include/generated/app_version.h
+    COMMAND ${CMAKE_COMMAND} -DZEPHYR_BASE=${ZEPHYR_BASE}
+      -DOUT_FILE=${PROJECT_BINARY_DIR}/include/generated/app_version.h
+      -DVERSION_TYPE=APP
+      -DVERSION_FILE=${APPLICATION_SOURCE_DIR}/VERSION
+      ${build_version_argument}
+      -P ${ZEPHYR_BASE}/cmake/gen_version_h.cmake
+    DEPENDS ${ZEPHYR_BASE}/VERSION ${git_dependency}
```
Comment (henrikbrixandersen) [inline CMakeLists.txt:564]: Should this be `${APPLICATION_SOURCE_DIR}/VERSION`?

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]

File: CMakeLists.txt (hunk 1 above)
Comment (tejlmand) [reply CMakeLists.txt:564]: of course it should.
Thanks for spotting

**Review by nordicjm — APPROVED** [review]

**Review by henrikbrixandersen — APPROVED** [review]

**Review by nashif — APPROVED** [review]

**Review by de-nordic — APPROVED** [review]

**Review by mbolivar-nordic — DISMISSED** [review]
This is application facing so it is end user facing. This means documentation is required.

**Review by carlescufi — APPROVED** [review]

**Comment by carlescufi** [comment]
> This is application facing so it is end user facing. This means documentation is required.

Can we split it up and provide the doc in a follow-up PR?

**Comment by mbolivar-nordic** [comment]
> Can we split it up and provide the doc in a follow-up PR? (We’ve done this before)

Sure -- @tejlmand?

**Comment by tejlmand** [comment]
> > Can we split it up and provide the doc in a follow-up PR? (We’ve done this before)
> 
> Sure -- @tejlmand?

Yep, let's do that.

**Comment by TomzBench** [comment]
Docs added yet?  couldn't find

**Comment by nordicjm** [comment]
Docs are pending review in https://github.com/zephyrproject-rtos/zephyr/pull/61068

> What happens if their are multiple apps in a same repo, they all get the same git-hash string, with independent version tracking? A "Manifest"/VERSION makes more sense if using git hash i think.

That should be explained in the doc, each application has its own VERSION file which is independent, only one define/variable would remain the same which would be the `git describe` one.

Different versions for different builds of the same application (i.e. with different configurations or for different boards) are impossible and unsupported, however.

**Comment by marc-hb** [comment]
> `#define KERNEL_BUILD_VERSION zephyr-v3.3.0-1081-gb6f97a492717`

So the build / version tweak is actually limited to one byte?
- #74357

I'm not sure, afraid I'm getting lost in the details.

**Pull request merged**